image/manifest_test.go

```diff
@@ -122,3 +103,45 @@ func TestUnpackLayer(t *testing.T) {
 		t.Fatal(err)
 	}
 }
+
+func createTarFile(name string, list []tarContent) (descriptor, error) {
+	f, err := os.Create(name)
+	if err != nil {
+		return descriptor{}, err
+	}
+	gw := gzip.NewWriter(f)
+	tw := tar.NewWriter(gw)
+
+	for _, content := range list {
+		if err = tw.WriteHeader(content.header); err != nil {
+			tw.Close()
+			gw.Close()
+			f.Close()
+			return descriptor{}, err
+		}
+		if _, err = io.Copy(tw, bytes.NewReader(content.b)); err != nil {
+			tw.Close()
+			gw.Close()
+			f.Close()
```

Comment:
Perhaps you avoid it because you want to ensure the file is completed and closed before you start hashing it.  You could jump through some hoops to get the `defer`ed functions to fire, but it's probably easier and more efficient to use something like:

```go
file, err := os.Create(name)  // + err checking
hash := sha256.New()
tarMultiWriter := io.MultiWriter(file, hash)
gzipWriter := gzip.NewWriter(tarMultiWriter)
```